Clarify use of E_DIAL_REFUSED in autonat (#430)
libp2p/specs: autonat/README.md

successfully, it sends a `DialResponse` with the `ResponseStatus` `OK`. It
SHOULD include the address it successfully dialed in its response.

If the receiver elects not to dial any addresses from the message - for example

> because they are invalid, because they resolve to the same host as the receiver,
> because they only include private addresses, because it does not support the

Reviewer: ```suggestion
because they are invalid, because it does not support the
```

> because they only include private addresses

In rust-libp2p we don't reject requests in this case. Directly after starting to listen, a peer won't know any of their external addresses and instead just their listening addresses (which are all private). The server still attempts to dial them, based on the address it observes this peer on, i.e. it replaces the IP of each address in the message with the observed IP. We do this anyway for DDoS protection (see https://github.com/status-im/nim-libp2p/pull/739#discussion_r917551966; @Tanguy suggested that we also add a section on for this, which imo would make sense).
Only if the observed address is private we may reject the peer (depending on whether this was set in our config or not).
> because they resolve to the same host as the receiver

Again, because we replace the IP with the observed one at least in rust-libp2p we don't really handle this case.

---
 Not sure how I think about having an exemplary list of rejection reasons. They seem rather go-lib2p2p specific right now. I'd propose to either properly specify in which cases a DialRequest shall be rejected, or keep it on a higher level to leave it up to the implementation (see suggestion).

> relevant transports, or because it has no capacity, is sends a `DialResponse`

Reviewer 2: ```suggestion
relevant transports, or because it has no capacity, it sends a `DialResponse`
```

> relevant transports, or because it has no capacity, is sends a `DialResponse`

Reviewer: ```suggestion
relevant transports, or because it has no capacity, it SHOULD send a `DialResponse`
```

> We should probably add that you can’t trust on your peer doing this correctly - the Go implementation has been sending the wrong error code for a long time (until Elena fixed it recently).

@marten-seemann What do you think about using the keyword _SHOULD_, in the sense that it is recommended but one can not trust that the other peer actually does it?

Reviewer 3: Technically, this is a MUST. You _should_ be able to trust your peer's reporting here.

In practice you can't. But that's just because the Go implementation was just not compliant with the spec, which brought us to the mess we're in right now (one more reason for Autonat v2!). I suggest adding a note here that this was the case.
